controllers/machinehealthcheck_targets.go

```diff
@@ -0,0 +1,257 @@
+	var nextCheckTimes []time.Duration
+	now := time.Now()
+
+	// machine has failed
+	if t.Machine.Status.FailureReason != nil {
+		logger.V(3).Info("Target is unhealthy", "reason", t.Machine.Status.FailureReason)
+		return true, time.Duration(0)
+	}
+
+	// the node has not been set yet
+	if t.Node == nil {
+		// status not updated yet
+		if t.Machine.Status.LastUpdated == nil {
+			return false, timeoutForMachineToHaveNode
+		}
+		if t.Machine.Status.LastUpdated.Add(timeoutForMachineToHaveNode).Before(now) {
+			logger.V(3).Info("Target is unhealthy: machine has no node", "duration", timeoutForMachineToHaveNode.String())
+			return true, time.Duration(0)
+		}
+		durationUnhealthy := now.Sub(t.Machine.Status.LastUpdated.Time)
+		nextCheck := timeoutForMachineToHaveNode - durationUnhealthy + time.Second
+		return false, nextCheck
+	}
+
+	// the node does not exist
+	if t.Node != nil && t.Node.UID == "" {
+		return true, time.Duration(0)
+	}
+
+	// check conditions
+	for _, c := range t.MHC.Spec.UnhealthyConditions {
+		nodeCondition := getNodeCondition(t.Node, c.Type)
+
+		// Skip when current node condition is different from the one reported
+		// in the MachineHealthCheck.
+		if nodeCondition == nil || nodeCondition.Status != c.Status {
+			continue
+		}
+
+		// If the condition has been in the unhealthy state for longer than the
+		// timeout, return true with no requeue time.
+		if nodeCondition.LastTransitionTime.Add(c.Timeout.Duration).Before(now) {
+			logger.V(3).Info("Target is unhealthy: condition is in state longer than allowed timeout", "condition", c.Type, "state", c.Status, "timeout", c.Timeout.Duration.String())
+			return true, time.Duration(0)
+		}
+
+		durationUnhealthy := now.Sub(nodeCondition.LastTransitionTime.Time)
+		nextCheck := c.Timeout.Duration - durationUnhealthy + time.Second
+		if nextCheck > 0 {
+			nextCheckTimes = append(nextCheckTimes, nextCheck)
+		}
+	}
+	return false, minDuration(nextCheckTimes)
+}
+
+// getTargetsFromMHC uses the MachineHealthCheck's selector to fetch machines
+// and their nodes targeted by the health check, ready for health checking.
+func (r *MachineHealthCheckReconciler) getTargetsFromMHC(clusterClient client.Client, cluster *clusterv1.Cluster, mhc *clusterv1.MachineHealthCheck) ([]healthCheckTarget, error) {
+	machines, err := r.getMachinesFromMHC(mhc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting machines from MachineHealthCheck")
+	}
+	if len(machines) == 0 {
+		return nil, nil
+	}
+
+	targets := []healthCheckTarget{}
+	for k := range machines {
+		target := healthCheckTarget{
+			MHC:     mhc,
+			Machine: &machines[k],
+		}
+		node, err := r.getNodeFromMachine(clusterClient, target.Machine)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, errors.Wrap(err, "error getting node")
+			}
+			// a node with only a name represents a
+			// not found node in the target
```

Comment:
This is subtle, + when combined with L100. Do you think it would be clearer if we added a `nodeMissing` field to `healthCheckTarget`?